test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc

```diff
@@ -0,0 +1,180 @@
+#include "test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h"
+
+#include <memory>
+
+#include "envoy/api/api.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PrivateKeyMethodProvider {
+
+int EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index = -1;
+
+void EcdsaPrivateKeyConnection::delayed_op() {
+  const std::chrono::milliseconds timeout_0ms{0};
+
+  timer_ = dispatcher_.createTimer([this]() -> void {
+    finished_ = true;
+    this->cb_.complete();
+    return;
+  });
+  timer_->enableTimer(timeout_0ms);
+}
+
+static ssl_private_key_result_t privateKeySign(SSL* ssl, uint8_t* out, size_t* out_len,
+                                               size_t max_out, uint16_t signature_algorithm,
+                                               const uint8_t* in, size_t in_len) {
+  (void)out;
+  (void)out_len;
+  (void)signature_algorithm;
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int hash_len;
+  EC_KEY* ec_key;
+  const EVP_MD* md;
+  bssl::ScopedEVP_MD_CTX ctx;
+  EcdsaPrivateKeyConnection* ops = static_cast<EcdsaPrivateKeyConnection*>(
+      SSL_get_ex_data(ssl, EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index));
+
+  if (!ops) {
+    return ssl_private_key_failure;
+  }
+
+  ec_key = ops->getPrivateKey();
+  if (!ec_key) {
+    return ssl_private_key_failure;
+  }
+
+  md = SSL_get_signature_algorithm_digest(signature_algorithm);
+  if (!md) {
+    EC_KEY_free(ec_key);
+    return ssl_private_key_failure;
+  }
+
+  // Calculate the digest for signing.
+  if (!EVP_DigestInit_ex(ctx.get(), md, nullptr) || !EVP_DigestUpdate(ctx.get(), in, in_len) ||
+      !EVP_DigestFinal_ex(ctx.get(), hash, &hash_len)) {
+    EC_KEY_free(ec_key);
+    return ssl_private_key_failure;
+  }
+
+  ops->out_ = static_cast<uint8_t*>(OPENSSL_malloc(max_out));
```

Comment:
You're using `OPENSSL_malloc` for ops&rarr;out_, but you don't need to because `privateKeyComplete` copies the value then frees it. So it could just be a std::vector or such to avoid the manual memory management. (I'm assuming that Envoy uses std::vector.)

Reply:
All right. I guess I was thinking the memory clearing would be good to have, but Vector looks nicer and leads to less code.